Привет, Юрий!) Меня зовут Дуолан👋 Поздравляю тебя с первым проектом по машинному обучению 🔥 
<br> Сегодня я буду проверять твой проект. Давай вместе доведем его до идеала😉
<br> Дальнейшее общение будет происходить на «ты», если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий («исправил», «не понятно как исправить ошибку», ...)
<br> Пожалуйста, не удаляй мои комментарии, они будут необходимы для повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет:</b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Замечание:</b> Если требуются исправления
</div>

Работа не может быть принята с красными комментариями.

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b>
    
привет, Дуолан 👋 
    
спасибо за твои проверку и советы! 
  
</div>


<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Открытие-файлов.-Общее-предоставление" data-toc-modified-id="Открытие-файлов.-Общее-предоставление-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Открытие файлов. Общее предоставление</a></span></li><li><span><a href="#Предварительный-анализ" data-toc-modified-id="Предварительный-анализ-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Предварительный анализ</a></span><ul class="toc-item"><li><span><a href="#Баланс-классов" data-toc-modified-id="Баланс-классов-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Баланс классов</a></span></li><li><span><a href="#Мультиколлинеарность" data-toc-modified-id="Мультиколлинеарность-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Мультиколлинеарность</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Деление-данных" data-toc-modified-id="Деление-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Деление данных</a></span><ul class="toc-item"><li><span><a href="#Определение-выборок" data-toc-modified-id="Определение-выборок-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Определение выборок</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Обучение-моделей" data-toc-modified-id="Обучение-моделей-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обучение моделей</a></span><ul class="toc-item"><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Анализ-качества-моделей" data-toc-modified-id="Анализ-качества-моделей-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Анализ качества моделей</a></span><ul class="toc-item"><li><span><a href="#Тестовая-выборка" data-toc-modified-id="Тестовая-выборка-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Тестовая выборка</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверка-модели-на-вменяемость" data-toc-modified-id="Проверка-модели-на-вменяемость-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверка модели на вменяемость</a></span><ul class="toc-item"><li><span><a href="#Константная-модель" data-toc-modified-id="Константная-модель-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Константная модель</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Модель для определения тарифа
Оператор мобильной связи «Мегалайн» выяснил: многие клиенты пользуются архивными тарифами. Оператор хочет построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра». В нашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы. 


**Цель исследования** 

Необходимо построить модель для задачи классификации, которая выберет подходящий тариф. 


**Ход исследования**

Предобработка данных не понадобится — это было выполнено раннее. Построим модель с максимально большим значением accuracy. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверим accuracy на тестовой выборке.


Таким образом, исследование пройдёт в четыре этапа:
 1. Обзор данных.
 2. Деление данных.
 3. Обучение моделей.
 4. Анализ качества моделей.
 5. Проверка модели на вменяемость. 

<div class="alert alert-block alert-success">
<b>Успех:</b> Отличное введение, важно чтобы в проектах было подробное описание: когда потенциальные работодатели будут смотреть твой проект в портфолио, им нужно будет понимать задачу, чтобы можно было оценить её решение :) Молодец, что описание пишешь своими словами, не копируешь дословно описание проекта для студентов, так оно не выглядит как инструкция, а работа выглядет более самостоятельной.
</div>

## Обзор данных

### Открытие файлов. Общее предоставление

Импортируем библиотеки для работы в Питоне.

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

<div class="alert alert-block alert-success">
<b>Успех:</b> Правильное решение — выделять отдельную ячейку в начале проекта под подключение библиотек и настройки. Tак код легче поддерживать.
</div>

Откроем файл с данными и изучим общую информацию. Прочитаем файл и сохраним его в датасет.

In [ ]:
try: 
    users = pd.read_csv('/datasets/users_behavior.csv')
except: 
    users = pd.read_csv('users_behavior.csv')

users

calls  minutes  messages   mb_used  is_ultra
0      40.0   311.90      83.0  19915.42         0
1      85.0   516.75      56.0  22696.96         0
2      77.0   467.66      86.0  21060.45         0
3     106.0   745.53      81.0   8437.39         1
4      66.0   418.74       1.0  14502.75         0
...     ...      ...       ...       ...       ...
3209  122.0   910.98      20.0  35124.90         1
3210   25.0   190.36       0.0   3275.61         0
3211   97.0   634.44      70.0  13974.06         0
3212   64.0   462.32      90.0  31239.78         0
3213   80.0   566.09       6.0  29480.52         1

[3214 rows x 5 columns]

Согласно документации к данным:
* сalls — количество звонков,
* minutes — суммарная длительность звонков в минутах,
* messages — количество sms-сообщений,
* mb_used — израсходованный интернет-трафик в Мб,
* is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

### Предварительный анализ

#### Баланс классов

Все столбцы, кроме is_ultra в наших выборках будем называть - признаки (features). Признак, который нужно предсказать, — целевой (target): в нашей задаче это колонка is_ultra.

Рассмотрим подробнее столбец "is_ultra". Проверим, как часто встречается класс «1» или «0». Количество уникальных значений подсчитаем методом value_counts().

In [ ]:
# frequencies of unique values

users['is_ultra'].value_counts(normalize=True)

0    0.693528
1    0.306472
Name: is_ultra, dtype: float64

Класс «1» встречается реже, доля составляет только 30.6%. Классы не сбалансированы, когда их соотношение далеко от 1:1.

#### Мультиколлинеарность

Мультиколлинеарность — корреляция независимых переменных, которая затрудняет оценку и анализ общего результата. Найдем коэффициент корреляции Пирсона — меру линейной связи в исходных данных.

In [ ]:
# calculate the correlation between the columns

users.corr()

calls   minutes  messages   mb_used  is_ultra
calls     1.000000  0.982083  0.177385  0.286442  0.207122
minutes   0.982083  1.000000  0.173110  0.280967  0.206955
messages  0.177385  0.173110  1.000000  0.195721  0.203830
mb_used   0.286442  0.280967  0.195721  1.000000  0.198568
is_ultra  0.207122  0.206955  0.203830  0.198568  1.000000

Можно сказать о наличии сильной корреляции между признаками "calls" и "minutes", коэффициент Пирсона для них близок к 1. Логичнее будет убрать "количество звонков", т.к. параметр "количество минут" трактуется более однозначно для тарификации, он точнее показывает длительность разговоров в минутах для пользователей. Результат запишем в новый датасет.

In [ ]:
# remove column 'calls'

users_upd = users.drop(['calls'], axis=1)

### Вывод

Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. У нас в распоряжении поведение 3214 пользователей. 

Предварительный анализ показал, что у предоставленных данных есть следующие особенности:
* Классы не сбалансированы в целевом признаке. Будущая модель будет угадывать "1", но их в данных меньшинство. Необходимо это учесть при определении метрик качества моделей.
* Была обнаружена мультиколлинеарность для признаков "количество звонков" и "количество минут". Столбец "calls" был удален. 

Результаты данного этапа позволят нам обучить модель быстрее и качественнее, а также выбрать более точные метрики для оценки ее работы.

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные на месте 👌
</div>

<div class="alert alert-block alert-info">
<b>Совет:</b>

Стоит посмотреть на соотношение классов в выборке. Это можно сделать так:

data.is_ultra.value_counts( )
    
Если есть дисбаланс, нужно смотреть на дополнительные метрики помимо accuracy. В данной выборке у нас только 985 наблюдений класса 1, тогда как остальные 2229 — класса 0. Представь, что модель предсказала правильно только 2229 наблюдений класса 0, accuracy будет около 70%. Вроде много, но по факту модель вообще не понимает и не умеет предсказывать класс 1. Такие нюансы можно оценить с помощью recall и precision.

Еще стоит проверить данные на наличие сильно-скоррелированных (мультиколлинеарных признаков). Если таковые есть, стоит удалить по одному из пары, иначе они будут доминировать при обучении.
    
У вас не было этого в тренажере, но мне хочется сделать ревью максимально полезным и развивающим :) Поэтому посмотри, пожалуйста, на указанные нюансы.

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b>
  
    
полезный совет! 👍
  
</div>


<div class="alert alert-block alert-warning">

<b>Изменения:</b>
    
* добавлен п.1.2
* обновил вывод

  
</div>


## Деление данных

### Определение выборок

Тестовой выборки нет. Значит, данные нужно разбить на три части: обучающую, валидационную и тестовую. Размеры тестового и валидационного наборов обычно равны. Исходные данные разбиваем в соотношении 3:1:1. Зафиксируем псевдослучайность для алгоритма обучения, укажем параметр random_state. Необходимо учесть также, что в целевом признаке присутствует дисбаланс. Укажем параметр stratify, который сохранит пропорции классов (наблюдения класса 0 и наблюдения класса 1) для целевого признака во всех трех выборках. 

In [ ]:
# definition of features and target
features = users_upd.drop('is_ultra', axis=1) 
target = users_upd['is_ultra']

# split data into train and valid sets 
features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.4, 
                                                                    random_state=123, stratify=target)

# split valid set into test set
features_valid, features_test, target_valid, target_test = train_test_split(features_valid, target_valid, test_size=0.5, 
                                                                  random_state=123, stratify=target_valid)

### Вывод

Мы подготовили три вида набора данных для создания будущей модели. Выборки имеют одинаковые пропорции классов (наблюдения класса 0 и наблюдения класса 1) для целевого признака.
* Обучающая выборка. 60% исходных данных. Выборка нужна для определения классификации и выявления взаимосвязей в признаках.
* Валидационная выборка. 20% исходных данных. Валидация показывает, как ведёт себя модель с незнакомыми данными и нет ли переобучения. Она также нужна, чтобы оценка качества была более надёжной.
* Тестовая выборка. 20% исходных данных. Новый набор данных с известными целевыми признаками нужен для проверки знаний, правильных предсказаний модели. Доля правильных предсказаний влияет на оценку качества accuracy.


<div class="alert alert-block alert-success">
<b>Успех:</b> Данные поделены верно, пропорции выбраны адекватные 👍
</div>

<div class="alert alert-block alert-info">
<b>Совет:</b>

Можно использовать параметр stratify = y в функции train_test_split( ), где y – соответствующая целевая переменная. Данные разделятся на два кластера (наблюдения класса 0 и наблюдения класса 1), и в валидационную выборку попадут по 20% наблюдений из каждого кластера:

![Screenshot%202021-09-18%20at%2017.31.53.png](attachment:Screenshot%202021-09-18%20at%2017.31.53.png)
    
То есть, если соотношение классов в оригинальной выборке было 2:5, то в обучающей и валидационной выборках оно сохранится таким же. Такой подход в чем-то решает проблему несбалансированных классов.

Это можно сделать так

Сначала нужно отделить от данных целевую переменную:

X = data.drop('is_ultra', axis=1)
y = data['is_ultra']

Потом делим данные на тренировочную и валидационную выборки:

X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.4, random_state=12345, stratify=y)

указываем stratify=y, чтобы сохранить соотношение классов при разбиении
    
Далее делим валидационную выборку на валидационную и тестовую:

X_valid, X_test, y_valid, y_test = train_test_split(X_valid, y_valid, test_size=0.5, random_state=12345, stratify=y_valid)

в этом случае stratify указываем y_valid</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b>
  
    
классный метод! 👍
  
</div>


<div class="alert alert-block alert-warning">

<b>Изменения:</b>
    
* изменил п.2.1
* обновил вывод

  
</div>


## Обучение моделей

Исследуем качество разных моделей, меняя гиперпараметры. 

### Решающее дерево

В нашем распоряжении обучающий набор данных и целевой признак, который нужно предсказать по остальным признакам, — выбор тарифа. Такие задачи относятся к классу «обучение с учителем». Воспользуемся методом DecisionTreeClassifier для классификации деревом решений. Необходимо перебрать в алгоритме обучения разные варианты глубины дерева, чтобы получить максимально высокое качество предсказания.

In [ ]:
# cycle for max_depth of DecisionTree

for depth in range(1, 10):
    model = DecisionTreeClassifier(max_depth=depth, random_state=123)
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid) 
    result = accuracy_score(target_valid, predictions)
    print(f'max_depth = {depth} : {result}')

max_depth = 1 : 0.7465007776049767
max_depth = 2 : 0.7822706065318819
max_depth = 3 : 0.7947122861586314
max_depth = 4 : 0.7838258164852255
max_depth = 5 : 0.7822706065318819
max_depth = 6 : 0.7838258164852255
max_depth = 7 : 0.7900466562986003
max_depth = 8 : 0.7853810264385692
max_depth = 9 : 0.7838258164852255


Модель лучше всего предсказывает с глубиной дерева 3. Accuracy наилучшей модели на валидационной выборке: 0.795.

### Случайный лес

Попробуем новый алгоритм классификации для модели — случайный лес. Случайный лес помогает улучшить результат предсказания и избежать переобучения. Необходимо перебрать в алгоритме обучения разные варианты глубины дерева и количества оценщиков, чтобы получить максимально высокое качество предсказания.

In [ ]:
# cycle for best result of RandomForest

from tqdm.notebook import tqdm

best_result = 0
best_depth = 0
best_est = 0

for est in tqdm(range(100, 1001, 100)):
    for depth in range(1, 22, 2):
        model = RandomForestClassifier(random_state=123, max_depth=depth, n_estimators=est)
        model.fit(features_train, target_train)
        result = model.score(features_valid, target_valid)
        if result > best_result:
            best_result = result
            best_depth = depth
            best_est = est
            
print(f'max_depth = {best_depth}, n_estimators = {best_est} : {best_result}')

  0%|          | 0/10 [00:00<?, ?it/s]

max_depth = 11, n_estimators = 100 : 0.8040435458786936


Модель лучше всего предсказывает с глубиной дерева 11, с минимальным числом оценщиков 100. Accuracy наилучшей модели на валидационной выборке: 0.804. 

<div class="alert alert-block alert-info">
<b>Совет:</b> Не бойся задавать широкий диапазон значений) Попробуй max_depth range(1, 22, 2), n_estimators range(100, 1001, 100). По умолчанию n_est равен 100
</div>

<div class="alert alert-block alert-warning">

<b>Изменения:</b>
    
* изменил п.3.1 и 3.2

  
</div>


### Логистическая регрессия

Попробуем ещё один алгоритм — логистическую регрессию.

In [ ]:
# accuracy of LogisticRegression

model = LogisticRegression(random_state=123)
model.fit(features_train, target_train)
score = model.score(features_valid, target_valid)
print(f'Accuracy модели на валидационной выборке: {score:.0%}')

Accuracy модели на валидационной выборке: 70%


### Вывод

Исследование качества разных моделей при изменении гиперпараметров дало следующие результаты:
* Решающее дерево. Accuracy наилучшей модели (глубина дерева 3) на валидационной выборке: 0.795.
* Случайный лес. Accuracy наилучшей модели (число оценщиков 200) на валидационной выборке: 0.804.
* Логистическая регрессия. Accuracy наилучшей модели на валидационной выборке: 0.70.

Модели, обученные методом Решающее дерево и Случайный лес, делают предсказание с лучшей точностью. Логистическая регрессия справляется с поставленной задачей хуже всего, доля правильных ответов меньше нашего критерия 0.75, этот метод не подходит для поставленной задачи.

<div class="alert alert-block alert-warning">

<b>Изменения:</b>
    
* обновил вывод

  
</div>


## Анализ качества моделей

### Тестовая выборка

Чтобы знать точно, как предсказывает модель, возьмём новый датасет — тестовый набор данных. Проверим, как модель при лучших гиперпараметрах справится с ним. Классы не сбалансированы, проверка только Accuracy не подходит. Исследуем наши методы с помощью различных метрик. Класс с меткой «1» называется положительным, с меткой «0» — отрицательным.
* Accuracy (англ. «правильность»). Отношение числа правильных ответов к размеру тестовой выборки.
* Precision (англ. «точность»). Определяет, как много отрицательных ответов нашла модель, пока искала положительные. Чем больше отрицательных, тем ниже точность.
* Recall (англ. «полнота»). Выявляет, какую долю положительных среди всех ответов выделила модель. Важно понимать, как хорошо модель их находит.

Решающее дерево:

In [ ]:
# accuracy, precision and recall for DecisionTree

model = DecisionTreeClassifier(max_depth=3, random_state=123)
model.fit(features_train, target_train)
predictions_train = model.predict(features_train) 
result_train = accuracy_score(target_train, predictions_train)
print(f'Accuracy модели на обучающей выборке: {result_train:.1%}')

predictions_valid = model.predict(features_valid) 
result_vaid = accuracy_score(target_valid, predictions_valid)
print(f'Accuracy модели на валидационной выборке: {result_vaid:.1%}')

predictions = model.predict(features_test) 
accuracy_tree = accuracy_score(target_test, predictions)
print(f'Accuracy модели на тестовой выборке: {accuracy_tree:.1%}')

recall_tree = recall_score(target_test, predictions)
print(f'Полнота модели на тестовой выборке: {recall_tree:.1%}')

precision_tree = precision_score(target_test, predictions)
print(f'Точность модели на тестовой выборке: {precision_tree:.1%}')

Accuracy модели на обучающей выборке: 79.7%
Accuracy модели на валидационной выборке: 79.5%
Accuracy модели на тестовой выборке: 80.2%
Полнота модели на тестовой выборке: 51.3%
Точность модели на тестовой выборке: 76.5%


Случайный лес:

In [ ]:
# accuracy, precision and recall for RandomForest

model = RandomForestClassifier(random_state=123, max_depth=11, n_estimators=100)
model.fit(features_train, target_train)

predictions_train = model.predict(features_train) 
result_train = accuracy_score(target_train, predictions_train)
print(f'Accuracy модели на обучающей выборке: {result_train:.1%}')

predictions_valid = model.predict(features_valid) 
result_vaid = accuracy_score(target_valid, predictions_valid)
print(f'Accuracy модели на валидационной выборке: {result_vaid:.1%}')

predictions = model.predict(features_test) 
accuracy_random = accuracy_score(target_test, predictions)
print(f'Accuracy модели на тестовой выборке: {accuracy_random:.1%}')

recall_random = recall_score(target_test, predictions)
print(f'Полнота модели на тестовой выборке: {recall_random:.1%}')

precision_random = precision_score(target_test, predictions)
print(f'Точность модели на тестовой выборке: {precision_random:.1%}')

Accuracy модели на обучающей выборке: 88.3%
Accuracy модели на валидационной выборке: 80.4%
Accuracy модели на тестовой выборке: 82.3%
Полнота модели на тестовой выборке: 57.4%
Точность модели на тестовой выборке: 79.0%


### Вывод

Модель, обученная методом случайный лес, делает предсказания лучше, решающее дерево имеет Accuracy на 2% хуже. Важно отметить, что полнота и точность у моделей далеки от хорошего результата. Лидирующий метод обучения имеет точность 79.0% при полноте 57,4%.

<div class="alert alert-block alert-success">
<b>Круто:</b> По модельной части замечаний нет 👌 Все сделано четко!) Молодец, что сопровождаешь каждый этап промежуточным выводом 👍
</div>

<div class="alert alert-block alert-warning">

<b>Изменения:</b>
    
* обновил весь раздел по твоим замечаниям. 👍

  
</div>


## Проверка модели на вменяемость

### Константная модель

Данные имеют дисбаланс классов для целевого признака. Класс «0» встречается почти в 70% случаев. Оценим адекватность модели, создадим константную модель: любому объекту она прогнозирует класс «0». Сохраним её предсказания в переменной target_pred_constant. Сравним полученные результаты с обученной моделью.

In [ ]:
# accuracy for constant model

const = 0
target_pred_constant = pd.Series(const, index=target_test.index)
result_const = accuracy_score(target_test, target_pred_constant)

print(f'Accuracy константной модели: {result_const:.1%}')

Accuracy константной модели: 69.4%


### Вывод

Accuracy модели, обученной методом случайный лес, больше, чем у константной модели. Выбранная модель прошла проверку на вменяемость.

<div class="alert alert-block alert-info">
<b>Совет:</b> Так как у нас наблюдается дисбаланс классов, то модель, дающая всегда отрицательные предсказания, будет иметь качество лучше, чем рандомно-предсказывающая модель
</div>

<div class="alert alert-block alert-warning">

<b>Изменения:</b>
    
* обновил весь раздел по твоим замечаниям. 👍

  
</div>


## Вывод

Были построены модели для задачи классификации по выбору подходящего тарифа пользователями. В распоряжении имелась информация о поведении 3214 пользователей за месяц. Анализ показал, что у предоставленных данных есть особенности: 
* дисбаланс классов (отношение 2:5) в целевом признаке. Проверка модели только на Accuracy не подходит, были выбраны дополнительные метрики качества: точность и полнота.
* Мультиколлинеарность признаков. Столбец "calls" был удален в целях облегчения модели, увеличения скорости ее тренировки и для наиболее точной оценки ее работы.

Исследование показало, что модель со случайным лесом делает предсказание на тестовой выборке с лучшей Accuracy 82%. Логистическая регрессия справляется с поставленной задачей хуже всего, доля правильных ответов меньше 0.75, для нашей модели не подходит. Модель со случайным лесом также успешно прошла проверку на вменяемость. Важно отметить, что полнота и точность у моделей далеки от хорошего результата. Лидирующий метод обучения имеет точность 79% при полноте 57.4%. Скорость работы не менее значимый критерий. Сделаем сводную таблицу по результатам исследования.

In [ ]:
pd.DataFrame({'модель':['Решающее дерево', 'Случайный лес'],
             'accuracy':[accuracy_tree, accuracy_random],
              'полнота':[recall_tree, recall_random],
              'точность':[precision_tree, precision_random],
             'скорость':['высокая', 'низкая']})

модель  accuracy   полнота  точность скорость
0  Решающее дерево  0.802488  0.512690  0.765152  высокая
1    Случайный лес  0.822706  0.573604  0.790210   низкая

<div class="alert alert-block alert-success">
<b>Успех:</b> Отличный результат!)
</div>

<div class="alert alert-block alert-warning">

<b>Изменения:</b>
    
* обновил весь раздел по твоим замечаниям. 👍

  
</div>


# <font color='orange'>Общее впечатление</font>
* Этот проект выполнен очень хорошо
* Достигнута необходимая метрика accuracy
* Видно, что ты уверенно кодишь и понимаешь базу машинного обучения
* Молодец, что структурируешь ноутбук, приятно проверять такие работы
* Проект можно уже принимать, но я хочу, чтобы ты внес небольшие коррективы:
1. В обучении Случайного леса добавить подбор гиперпараметра max_depth
2. Посмотреть на скор константной модели, которая всегда предсказывает один класс
* Также я оставил несколько советов. Исправь их если останется время.
* Жду новую версию проекта 👋

# <font color='orange'>Общее впечатление 2</font>
* Cпасибо за быстрое исправление
* Теперь проект выглядит лучше
* Критических замечаний нет
* Молодец, отличная работа 👍
* Надеюсь, ревью было полезным
* Удачи в дальнейшем обучении👋

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b>
  
    
огромное спасибо за твои советы и комментарии! это помогло мне лучше понять задачи МО и сделать хороший проект 🤓😎
    
Надеюсь, я все корректно исправил.
  
</div>


## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
